packages/taiko-client/driver/preconf_blocks/server.go

```diff
@@ -1074,59 +1083,68 @@ func (s *PreconfBlockAPIServer) TryImportingPayload(
-			args       = &miner.BuildPayloadArgs{
-				Parent:       parentInFork.ParentHash(),
-				Timestamp:    parentInFork.NumberU64(),
-				FeeRecipient: parentInFork.Coinbase(),
-				Random:       parentInFork.MixDigest(),
+			payloadID engine.PayloadID
+			parentID  *big.Int
+		)
+
+		if cachedParent != nil {
+			if len(cachedParent.Payload.Transactions) == 0 {
+				return false, fmt.Errorf("cached parent envelope has empty transactions: %s", msg.ExecutionPayload.ParentHash.Hex())
+			}
+			decompressedTxs, err := utils.Decompress(cachedParent.Payload.Transactions[0])
+			if err != nil {
+				return false, fmt.Errorf("failed to decompress cached parent tx list: %w", err)
+			}
+			txListHash := crypto.Keccak256Hash(decompressedTxs)
+			args := &miner.BuildPayloadArgs{
+				Parent:       cachedParent.Payload.ParentHash,
+				Timestamp:    uint64(cachedParent.Payload.Timestamp),
+				FeeRecipient: cachedParent.Payload.FeeRecipient,
+				Random:       common.Hash(cachedParent.Payload.PrevRandao),
 				Withdrawals:  make([]*types.Withdrawal, 0),
 				Version:      engine.PayloadV2,
 				TxListHash:   &txListHash,
 			}
-		)
+			payloadID = args.Id()
+			parentID = new(big.Int).SetUint64(uint64(cachedParent.Payload.BlockNumber))
 
-		payloadID := args.Id()
-		var sig [65]byte
-		if msg.Signature != nil {
-			sig = *msg.Signature
-		}
-		// update L1 Origin if the parent block is in the fork chain, we are building
-		// on an orphaned block.
-		_, err = s.rpc.L2Engine.UpdateL1Origin(ctx, &rawdb.L1Origin{
-			BuildPayloadArgsID: payloadID,
-			BlockID:            parentInFork.Number(),
-			L2BlockHash:        msg.ExecutionPayload.ParentHash,
-			IsForcedInclusion:  msg.IsForcedInclusion != nil && *msg.IsForcedInclusion,
-			Signature:          sig,
-		})
-		if err != nil {
-			return false, fmt.Errorf("failed to update L1 origin: %w", err)
-		}
+			var sig [65]byte
+			if msg.Signature != nil {
+				sig = *msg.Signature
+			}
 
-		log.Info("Updated L1 Origin for the parent block in the fork chain",
-			"peer", from,
-			"blockID", uint64(msg.ExecutionPayload.BlockNumber),
-			"hash", msg.ExecutionPayload.BlockHash.Hex(),
-			"signature", common.Bytes2Hex(sig[:]),
-		)
+			// Update L1 Origin for the parent block via cached data.
+			if _, err := s.rpc.L2Engine.UpdateL1Origin(ctx, &rawdb.L1Origin{
+				BuildPayloadArgsID: payloadID,
+				BlockID:            parentID,
+				L2BlockHash:        msg.ExecutionPayload.ParentHash,
+				IsForcedInclusion:  msg.IsForcedInclusion != nil && *msg.IsForcedInclusion,
+				Signature:          sig,
+			}); err != nil {
+				return false, fmt.Errorf("failed to update L1 origin: %w", err)
+			}
+
+			log.Info("Updated L1 Origin for orphan parent via cache",
+				"peer", from,
+				"parentID", parentID,
+				"parentHash", msg.ExecutionPayload.ParentHash.Hex(),
+				"signature", common.Bytes2Hex(sig[:]),
+			)
+		} else {
+			log.Info("Orphan parent detected but not found in cache; skipping L1Origin update",
```

Comment:
logically if we cache all correctly I believe this else should never be hit bc of the max blocks per proposal param, and an outdated L1Origin will trigger a reorg on proposal, but its good to have regardless. I would change it to a log.Warn though